composite_broker/src/broker.rs

```diff
@@ -0,0 +1,298 @@
+            }
+        } // end new
+
+        // accept a connect packet
+        // param: address of the client, connect packet
+        // return: connect ack
+        pub fn accept_connect(&mut self, addr: String, _connect_packet: Connect) -> Message {
+            let addr_port: Vec<&str> = addr.split(":").collect();
+            // error checking that client isn't already connected
+            if self.find_client_by_address(addr_port[0], addr_port[1]) {
+                let packet = Message::ConnAck(ConnAck { 
+                    code: ReturnCode::Rejected(RejectedReason::Congestion)
+                });
+                return packet
+            }
+            // make a new client 
+            self.num_clients += 1;
+            let c: Client = Client::new(self.num_clients.to_string(), addr_port[0].to_string(), addr_port[1].to_string());
+
+            // add to client list
+            self.client_list.push(c);
+
+            // create connect_ack packet
+            let conn_ack = Message::ConnAck(ConnAck { 
+                code: ReturnCode::Accepted 
+            } );
+            self.num_packets +=1;
+            conn_ack
+        } // end connect
+
+
+        // accept a subscribe packet -> (Packet, Option<Vec<String>>)
+        #[allow(unused_assignments)]
+        pub fn accept_sub(&mut self, addr: String, packet: Subscribe) -> Message {
+            self.num_packets +=1;
+            let addr_port: Vec<&str> = addr.split(":").collect();
+            let mut cli: Client = Client::new(String::new(), String::new(), String::new());
+            // find the client in reference
+            for client in self.client_list.clone() {
+                if client.address.eq(&addr_port[0].to_string()) {
+                    cli = client.clone();
+                } 
+            }
+
+            // if not found
+            if cli.client_id.trim().is_empty() {
+                println!("\nClient not found");
+                let sub_ack = Message::SubAck(SubAck {
+                    flags: packet.flags,
+                    msg_id: self.num_packets,
+                    topic_id: match packet.topic {
+                        mqtt_sn::TopicNameOrId::Name(_) => 0,
+                        mqtt_sn::TopicNameOrId::Id(id) => id
+                    },
+                    code: ReturnCode::Rejected(RejectedReason::NotSupported)
+                });
+                
+                return sub_ack;
+            }
+
+            // check if topic is in list
+            let mut topic = 0;
+            match packet.topic {
+                mqtt_sn::TopicNameOrId::Id(id) => {
+                    // println!("Found id: {}", id);
+                    topic = id
+                },
+                mqtt_sn::TopicNameOrId::Name(name) => {
+                    // generate a random number that is not a key in the list already
+                    // println!("Topic is : {}", name.to_string());
+                    let mut rng = rand::thread_rng();
```

Comment:
This can be *very* expensive. Might just want to use an ever-increasing number that is a `u64` so that it never practically wraps around? For example, https://docs.rs/atomic-counter/latest/atomic_counter/